### PR TITLE
rubocop: Fix some TODOs for `Style/` cops

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

- Part of https://github.com/Homebrew/brew/issues/15297.
- The rules with <= 100 offenses that made sense to me and/or were autocorrectable.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
All make sense to me too, thanks @issyl0!

**Comment by Bo98** [comment]
The `Style/YodaExpression` fixes seem odd to me to be honest, and feels weird to enforce that on taps. In fact, it appears that doing those corrections is actually introducing bugs as the order of the expressions on some things (e.g. Pathname) matters greatly. I don't think we should enable that cop.

LGTM otherwise.

**Comment by Bo98** [comment]
Actually I have some problems with the other ones:

`Style/SingleLineBlockParams` - this is unsafe and incorrect in some cases (see `system_command` change):

```
irb(main):001:0> [[1,2],[3,4]].select { |a| a == 1 }
=> []
irb(main):002:0> [[1,2],[3,4]].select { |a,| a == 1 }
=> [[1, 2]]
```

`Style/ArrayCoercion` - this is explicitly documented as containing false positives.

`Style/SingleLineBlockParams` - this makes all single line blocks have generic parameter names `acc` and `elem`. Looking at the fix for it in this PR, I do not see that being an improvement at all.

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -956,7 +956,7 @@ def bottle_prefix
   # The directory where the formula's installation or test logs will be written.
   # @private
   def logs
-    HOMEBREW_LOGS + name
+    name + HOMEBREW_LOGS
```
Comment (carlocab) [inline Library/Homebrew/formula.rb:959]: This is not equivalent.

File: Library/Homebrew/keg.rb [hunk 2]
```diff
@@ -290,7 +290,7 @@ def unlink(verbose: false, dry_run: false)
                                           .select(&:exist?)
     keg_directories.each do |dir|
       dir.find do |src|
-        dst = HOMEBREW_PREFIX + src.relative_path_from(path)
+        dst = src.relative_path_from(path) + HOMEBREW_PREFIX
```
Comment (carlocab) [inline Library/Homebrew/keg.rb:293]: Not equivalent.

File: Library/Homebrew/keg.rb [hunk 3]
```diff
@@ -624,7 +624,7 @@ def link_dir(relative_dir, verbose: false, dry_run: false, overwrite: false)
     root.find do |src|
       next if src == root
 
-      dst = HOMEBREW_PREFIX + src.relative_path_from(path)
+      dst = src.relative_path_from(path) + HOMEBREW_PREFIX
```
Comment (carlocab) [inline Library/Homebrew/keg.rb:627]: Also not equivalent.

**Review by carlocab — CHANGES_REQUESTED** [review]
I agree with @Bo98, and probably feel even more strongly.

**Comment by issyl0** [comment]
Thanks for your considered, detailed feedback! ❤️

I had it on my list to revisit the failures here and figure out what to do about them - that may be to disable the cops again.

**Comment by MikeMcQuaid** [comment]
> Some of them here have very good reasons. I'm personally against `EnabledByDefault` as a result (it greatly increases the risk of RuboCop updates causing bugs from `brew sytle --fix`, as demonstrated here) but I realise I'm late in expressing my view on that.

Some of them have good reasons, some of them do not. Given we have humans in the loops before these changes are made: I think this process has worked exactly how we hoped it would.

**Comment by MikeMcQuaid** [comment]
> There has been a trust built that `brew style --fix` won't change code behaviour - as shown by us not running full CI whenever we apply it to Homebrew/core.

I'm not sure whether `brew style --fix` is only running "safe" autocorrections but, if it's not, it probably should start doing so. If we do that and we're only manually making the non-safe ones: that seems fine to me, too.

File: Library/Homebrew/formula.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:959]: `HOMEBREW_LOGS/name` is probably more appropriate here and won't be changed in this way. Arguably this is a poor use of the `+` operator.

**Comment by Bo98** [comment]
> > There has been a trust built that `brew style --fix` won't change code behaviour - as shown by us not running full CI whenever we apply it to Homebrew/core.
> 
> I'm not sure whether `brew style --fix` is only running "safe" autocorrections but, if it's not, it probably should start doing so. If we do that and we're only manually making the non-safe ones: that seems fine to me, too.

Problem is some of these are a level further: it's not just that the fix is unsafe but that the entire cop itself is not reliable and following the manual instructions would equally lead to bad code.

> Some of them have good reasons, some of them do not. Given we have humans in the loops before these changes are made: I think this process has worked exactly how we hoped it would.

How many cops have we gained from the process? The bad rate from the ones in this PR is already >50%, and these were supposed to be the simple ones.

I ask because I see no reason why we can't do the inverse. We already do so for rubocop-rails.

**Comment by MikeMcQuaid** [comment]
> Problem is some of these are a level further: it's not just that the fix is unsafe but that the entire cop itself is not reliable and following the cop to apply a manual fix could equally lead to bad code.

Sure, in those cases: we disable the cop.

> How many cops have we gained from the process? The bad rate from the ones in this PR is already >50%, and these were supposed to be the simple ones.

This work is still in progress. My recent experience on a non-Homebrew project has found this to be a valuable change.

> I ask because I see no reason why we can't do the inverse, given we already do so for rubocop-rails.

Rubocop Rails is actively incorrect in many cases because we're not a Rails project. We're only using it for the very small parts of ActiveSupport we use.

Of course there's no reason why we can't do the inverse, I just don't think we should.

**Pull request closed without merging**